compiler/src/iree/compiler/Codegen/SPIRV/KernelConfig.cpp

```diff
@@ -131,74 +131,101 @@ LogicalResult setConvOpConfig(linalg::LinalgOp linalgOp,
+  const bool isNHWC = isa<linalg::Conv2DNhwcHwcfOp>(*linalgOp) ||
+                      isa<linalg::DepthwiseConv2DNhwcHwcOp>(*linalgOp);
+  if (!isNCHW & !isNHWC) return success();
+
+  const int icIndex = isNHWC ? 3 : 1;
+  const int ohIndex = isNHWC ? 1 : 2;
+  const int owIndex = isNHWC ? 2 : 3;
+  const int ocIndex = isNHWC ? 3 : 1;
+
+  if (ShapedType::isDynamic(inputShape[icIndex]) ||
+      llvm::any_of(outputShape.drop_front(), ShapedType::isDynamic)) {
     return success();
   }
 
-  const int64_t ic = inputShape[3];
-  const int64_t oh = outputShape[1], ow = outputShape[2], oc = outputShape[3];
+  const int64_t ic = inputShape[icIndex], oc = outputShape[ocIndex];
+  const int64_t oh = outputShape[ohIndex], ow = outputShape[owIndex];
 
   // The conversion pipeline requires the input channel dimension to be some
   // multipler of four, or less than four.
   if (!(ic % 4 == 0 || ic < 4)) return success();
 
-  // The core idea is to distribute the convolution OH/OW/OC dimension to the
-  // workgroup Z/Y/X dimension, with each thread in a workgroup handling
-  // multiple vector elements. We try to 1) utilize all threads in a subgroup,
-  // and 2) handle an optimal tile size along each dimension.
+  // The core idea is to distribute the convolution dimensions to the workgroup
+  // Z/Y/X dimensions, with each thread in a workgroup handling multiple vector
+  // elements. We try to 1) utilize all threads in a subgroup, and 2) handle an
+  // optimal tile size along each dimension.
 
   int64_t residualThreads = subgroupSize;
   int64_t residualTilingFactor = bestTilingFactor;
 
-  SmallVector<int64_t, 3> workgroupSize(3, 1);    // (X, Y, Z)
-  SmallVector<int64_t> workgroupTileSizes(4, 0);  // (N, OH, OW, OC)
-  SmallVector<int64_t> threadTileSizes(4, 0);     // (N, OH, OW, OC)
-
-  // Deduce the configuration for the OC dimension.
-  if (!distributeToOneDim(oc, /*isInnerMostDim=*/true, residualThreads,
-                          residualTilingFactor, workgroupSize[0],
-                          workgroupTileSizes[3], threadTileSizes[3]))
-    return success();
-
-  // Deduce the configruation for the OW and OH dimension. Try to make them even
-  // if possible given we typically have images with the same height and width.
-  const bool tileToSquare = distributeToSquare(
-      oh, ow, residualThreads, residualTilingFactor,
-      llvm::makeMutableArrayRef(workgroupSize).drop_front(),
-      llvm::makeMutableArrayRef(workgroupTileSizes).drop_front().drop_back(),
-      llvm::makeMutableArrayRef(threadTileSizes).drop_front().drop_back());
-
-  // Otherwise treat OW and OH separately to allow them to have different number
-  // of threads and tiling size.
-  if (!tileToSquare) {
-    if (!distributeToOneDim(ow, /*isInnerMostDim=*/false, residualThreads,
+  SmallVector<int64_t, 3> workgroupSize(3, 1);  // (X, Y, Z)
+  SmallVector<int64_t> workgroupTileSizes(4, 0);
+  SmallVector<int64_t> threadTileSizes(4, 0);
+
+  if (isNCHW) {
```

Comment:
nit: would it be possible to split this function into a couple smaller ones? I don't have any concrete issues with it, but it's getting quite long

Reply:
I did a refactoring prior to this to extract quite some logic out https://github.com/iree-org/iree/pull/10571. Further splitting would mean duplicating quite some common code. So I'd keep it for now.